src/AzureAppConfigurationImpl.ts

```diff
@@ -55,23 +83,80 @@ export class AzureAppConfigurationImpl extends Map<string, unknown> implements A
                 listOptions.requestOptions = {
-                    customHeaders: this.customHeaders()
+                    customHeaders: this.customHeaders(requestType)
                 }
             }
 
             const settings = this.client.listConfigurationSettings(listOptions);
 
             for await (const setting of settings) {
                 if (setting.key) {
-                    const [key, value] = await this.processAdapters(setting);
-                    const trimmedKey = this.keyWithPrefixesTrimmed(key);
-                    keyValues.push([trimmedKey, value]);
+                    const keyValuePair = await this.processKeyValues(setting);
+                    keyValues.push(keyValuePair);
+                }
+                // update etag of sentinels
+                const matchedSentinel = this.sentinels?.find(s => s.key === setting.key && (s.label ?? null) === setting.label); // Workaround: as undefined label represents the same with null.
+                if (matchedSentinel) {
+                    matchedSentinel.etag = setting.etag;
                 }
             }
         }
         for (const [k, v] of keyValues) {
             this.set(k, v);
         }
+        this.lastUpdateTimestamp = Date.now();
+    }
+
+    public async refresh(): Promise<void> {
+        // if no refreshOptions set, return
+        if (this.sentinels === undefined || this.sentinels.length === 0 || this.refreshIntervalInMs === undefined) {
+            return Promise.resolve();
+        }
+        // if still within refresh interval, return
+        const now = Date.now();
+        if (now < this.lastUpdateTimestamp + this.refreshIntervalInMs) {
+            return Promise.resolve();
+        }
+
+        // try refresh if any of watched settings is changed.
+        let needRefresh = false;
+        for (const sentinel of this.sentinels) {
+            const response = await this.client.getConfigurationSetting(sentinel, {
+                onlyIfChanged: true
+                // TODO: do we trace this request by adding custom headers?
+            });
+            if (response.statusCode !== 304) { // TODO: can be more robust, e.g. === 200?
+                // sentinel changed.
+                sentinel.etag = response.etag;// update etag of the sentinel
+                needRefresh = true;
+                break;
+            }
+        }
+        if (needRefresh) {
+            await this.load(RequestType.Watch);
+            // run callbacks in async
+            if (this.onRefreshListeners !== undefined) {
+                for (const listener of this.onRefreshListeners) {
+                    listener();
+                }
+            }
+        }
+    }
+
+    public onRefresh(listener: () => any, thisArg?: any): Disposable {
+        if (this.onRefreshListeners === undefined) {
+            // TODO: Add unit tests
+            throw new Error("Illegal operation because refreshOptions is not provided on loading.");

```

Comment:
I see `onRefreshListeners` is initialized in contructor. Is there a code path that it's not initialized ?

Reply:
It's initialized as a linkedList only when refreshOptions is specified. So if no refreshOptions provided, it's undefined. 

An alternative impl is we always initialize it as an empty list, if we want to get rid of the null checks.

Reply:
I will always init refreshListener as an emtpy list. And use refreshOptions.enabled to determine whether refresh is enabled and the onRefresh API is available.

Reply:
updated.